# [Tarea 03] Ejercicios Unidad 01 - Algoritmos y Convergencia

- Dax Navarrete
https://github.com/daxrpm/numerical-methods_homeworks

## CONJUNTO DE EJERCICIOS 1.3

Resuelva los siguientes ejercicios, tome en cuenta que debe mostrar el desarrollo completo.


### 1. Aritmética de corte de tres dígitos

Utilice aritmética de corte de tres dígitos para calcular las siguientes sumas. Para cada parte, ¿qué método es más preciso y por qué?


#### a) $ \displaystyle \sum_{i=1}^{10} \frac{1}{i^2} $

Primero por $ 1 + \frac{1}{4} + \cdots + \frac{1}{100} $, y luego por $ \frac{1}{100} + \frac{1}{81} + \cdots + 1 $


**Demostración de funciones auxiliares:**


In [ ]:
def sum_for(iterations: int, i: int):
    result = 0
    for a in range(i, iterations + 1):
        print(f"1/{a**2}", end="  ")
        result += 1/a**2
    return result

In [2]:
def sum_back(iterations: int, i: int):
    result = 0
    for a in range(iterations, i-1, -1):
        print(f"1/{a**2}", end="  ")
        result += 1/a**2
    return result

In [3]:
sum_for(10, 1)

1/1  1/4  1/9  1/16  1/25  1/36  1/49  1/64  1/81  1/100  

1.5497677311665408

In [4]:
sum_back(10, 1)

1/100  1/81  1/64  1/49  1/36  1/25  1/16  1/9  1/4  1/1  

1.5497677311665408

#### Resultados del Ejercicio 1

**Parte (a): $ \displaystyle \sum_{i=1}^{10} \frac{1}{i^2} $**

Suma de menor a mayor (ascendente): $ 1.53 $

Suma de mayor a menor (descendente): $ 1.54 $

**Parte (b): $ \displaystyle \sum_{i=1}^{10} \frac{1}{i^3} $**

Suma de menor a mayor (ascendente): $ 1.16 $

Suma de mayor a menor (descendente): $ 1.19 $

**Análisis:**

El método de suma de mayor a menor (descendente) es más preciso en ambos casos. Esto se debe a que al comenzar con los términos más pequeños, se minimiza la pérdida de precisión que ocurre cuando se suman números de magnitudes muy diferentes. Al usar aritmética de corte de tres dígitos, sumar un número pequeño a uno grande puede resultar en la pérdida del número pequeño debido al redondeo.


#### b) $ \displaystyle \sum_{i=1}^{10} \frac{1}{i^3} $

Primero por $ 1 + \frac{1}{8} + \frac{1}{27} + \cdots + \frac{1}{1000} $, y luego por $ \frac{1}{1000} + \frac{1}{729} + \cdots + 1 $

**Implementación con aritmética de corte de tres dígitos:**


In [7]:
def chop(x, d=3):
    if x == 0:
        return 0.0
    from math import log10, floor
    exp = floor(log10(abs(x)))
    mant = abs(x) / (10**exp)
    mant = int(mant * (10**(d-1))) / (10**(d-1)) 
    return (mant * (10**exp)) if x > 0 else -(mant * (10**exp))


def sum_chop_forward(terms):
    s = 0.0
    for t in terms:
        s = chop(s + t)
    return s


def sum_chop_backward(terms):
    s = 0.0
    for t in reversed(terms):
        s = chop(s + t)
    return s


terms_a = [1/(i**2) for i in range(1,11)]

terms_b = [1/(i**3) for i in range(1,11)]

print("=== Parte (a): Σ 1/i^2 ===")
print("Suma de menor a mayor:", sum_chop_forward(terms_a))
print("Suma de mayor a menor:", sum_chop_backward(terms_a))

print("\n=== Parte (b): Σ 1/i^3 ===")
print("Suma de menor a mayor:", sum_chop_forward(terms_b))
print("Suma de mayor a menor:", sum_chop_backward(terms_b))


=== Parte (a): Σ 1/i^2 ===
Suma de menor a mayor: 1.53
Suma de mayor a menor: 1.54

=== Parte (b): Σ 1/i^3 ===
Suma de menor a mayor: 1.16
Suma de mayor a menor: 1.19


### 2. Serie de Maclaurin para la función arcotangente

La serie de Maclaurin para la función arcotangente converge para $ -1 < x \leq 1 $ y está dada por:

$$
\arctan x = \lim_{n \to \infty} P_n(x) = \lim_{n \to \infty} \sum_{i=1}^{n} \frac{(-1)^{i+1} x^{2i-1}}{2i-1}
$$


#### a) Determinación del número de términos para $ |4P_n(1) - \pi| < 10^{-3} $

Utilice el hecho de que $ \tan \frac{\pi}{4} = 1 $ para determinar el número $ n $ de términos de la serie que se necesita sumar para garantizar que $ |4P_n(1) - \pi| < 10^{-3} $.

**Desarrollo:**

Se usa la serie:

$$
\arctan(1)=\frac{\pi}{4}=\sum_{i=1}^{\infty}\frac{(-1)^{i+1}}{2i-1}
$$

Suma parcial:

$$
P_n(1)=\sum_{i=1}^{n}\frac{(-1)^{i+1}}{2i-1}
$$

Error de serie alternante:

$$
\left|\frac{\pi}{4}-P_n(1)\right|\le \frac{1}{2n+1}
$$

Error en $ \pi $:

$$
|\pi-4P_n(1)| \le \frac{4}{2n+1}
$$

Condición requerida:

$$
\frac{4}{2n+1}<10^{-3}
$$

Resolviendo:

$$
2n+1>4000
$$

$$
2n>3999
$$

$$
n>1999.5
$$

Por lo tanto:

$$
n=2000
$$

Verificación:

$$
\frac{4}{2(2000)+1}=\frac{4}{4001} \approx 0.0009997 < 10^{-3}
$$

**Resultado:**  
$$
n = 2000 \text{ términos}
$$


In [16]:
import math

def calcular_arcotangente(n: int) -> float:

    return sum(((-1) ** (i + 1)) / (2 * i - 1) for i in range(1, n + 1))

def termos_necesarios_para_pi(tol: float = 1e-3) -> tuple:
    
    if tol <= 0:
        raise ValueError("tol debe ser mayor que 0")
    n = max(1, math.ceil((4 / tol - 1) / 2))
    while True:
        p = calcular_arcotangente(n)
        err = abs(4 * p - math.pi)
        if err < tol:
            return n, err
        n += 1


n, valor_encontrado = termos_necesarios_para_pi(1e-3)
arcotangente = calcular_arcotangente(n)
n, valor_encontrado

(2000, 0.0004999999687496093)

#### b) Determinación del número de términos para $ |\pi - 4P_n(1)| < 10^{-10} $

El lenguaje de programación C++ requiere que el valor de $ \pi $ se encuentre dentro de $ 10^{-10} $. ¿Cuántos términos de la serie se necesitarían sumar para obtener este grado de precisión?

**Desarrollo:**

Usando la misma serie alternante, necesitamos:

$$
\frac{4}{2n+1} < 10^{-10}
$$

$$
2n+1 > 4 \times 10^{10}
$$

$$
n > \frac{4 \times 10^{10} - 1}{2} \approx 2 \times 10^{10}
$$

**Implementación en C++:**

   ~/Desktop/EPN/4to/metodos_numericos/hw_2 on    master ?2 ❯ nvim test_sum.cpp                                                                                       hw_2

   ~/Desktop/EPN/4to/metodos_numericos/hw_2 on    master ?3 ❯ g++ test_sum.cpp -o test_sum                                                                            hw_2

   ~/Desktop/EPN/4to/metodos_numericos/hw_2 on    master ?4 ❯ ./test_sum                                                                                              hw_2

n = 100000000  error = 1.00005e-08
n = 200000000  error = 5.00039e-09
n = 300000000  error = 3.33377e-09
n = 400000000  error = 2.50073e-09
n = 500000000  error = 2.00054e-09
n = 600000000  error = 1.66745e-09
n = 700000000  error = 1.42973e-09
n = 800000000  error = 1.25149e-09
n = 900000000  error = 1.11273e-09
n = 1000000000  error = 1.00174e-09
n = 1100000000  error = 9.10811e-10
n = 1200000000  error = 8.34939e-10
n = 1300000000  error = 7.70853e-10
n = 1400000000  error = 7.16047e-10
n = 1500000000  error = 6.68369e-10
n = 1600000000  error = 6.26764e-10
n = 1700000000  error = 5.90033e-10
n = 1800000000  error = 5.57271e-10
n = 1900000000  error = 5.28055e-10
n = 2000000000  error = 5.01716e-10
n = 2100000000  error = 4.77784e-10
n = 2200000000  error = 4.56089e-10
n = 2300000000  error = 4.36348e-10
n = 2400000000  error = 4.1815e-10
n = 2500000000  error = 4.01522e-10
n = 2600000000  error = 3.86124e-10
n = 2700000000  error = 3.71824e-10
n = 2800000000  error = 3.58631e-10
n = 2900000000  error = 3.46263e-10
n = 3000000000  error = 3.34868e-10
n = 3100000000  error = 3.24112e-10
n = 3200000000  error = 3.1407e-10
n = 3300000000  error = 3.04596e-10
n = 3400000000  error = 2.95735e-10
n = 3500000000  error = 2.87419e-10
n = 3600000000  error = 2.79518e-10
n = 3700000000  error = 2.72062e-10
n = 3800000000  error = 2.64918e-10
n = 3900000000  error = 2.58204e-10
n = 4000000000  error = 2.51854e-10
n = 4100000000  error = 2.45681e-10
n = 4200000000  error = 2.39828e-10
n = 4300000000  error = 2.34243e-10
n = 4400000000  error = 2.28977e-10
n = 4500000000  error = 2.23928e-10
n = 4600000000  error = 2.19099e-10
n = 4700000000  error = 2.14441e-10
n = 4800000000  error = 2.09997e-10
n = 4900000000  error = 2.0576e-10
n = 5000000000  error = 2.01592e-10
n = 5100000000  error = 1.97599e-10
n = 5200000000  error = 1.93847e-10
n = 5300000000  error = 1.90211e-10
n = 5400000000  error = 1.86684e-10
n = 5500000000  error = 1.83331e-10
n = 5600000000  error = 1.80103e-10
n = 5700000000  error = 1.76978e-10
n = 5800000000  error = 1.73968e-10
n = 5900000000  error = 1.71037e-10
n = 6000000000  error = 1.68218e-10
n = 6100000000  error = 1.65482e-10
n = 6200000000  error = 1.62874e-10
n = 6300000000  error = 1.60313e-10
n = 6400000000  error = 1.57832e-10
n = 6500000000  error = 1.55411e-10
n = 6600000000  error = 1.53089e-10
n = 6700000000  error = 1.50834e-10
n = 6800000000  error = 1.48639e-10
n = 6900000000  error = 1.46554e-10
n = 7000000000  error = 1.44449e-10
n = 7100000000  error = 1.42416e-10
n = 7200000000  error = 1.40458e-10
n = 7300000000  error = 1.38528e-10
n = 7400000000  error = 1.3666e-10
n = 7500000000  error = 1.34856e-10
n = 7600000000  error = 1.33083e-10
n = 7700000000  error = 1.3139e-10
n = 7800000000  error = 1.29728e-10
n = 7900000000  error = 1.28118e-10
n = 8000000000  error = 1.26533e-10
n = 8100000000  error = 1.24992e-10
n = 8200000000  error = 1.23484e-10
n = 8300000000  error = 1.21997e-10
n = 8400000000  error = 1.20563e-10
n = 8500000000  error = 1.19156e-10
n = 8600000000  error = 1.17784e-10
n = 8700000000  error = 1.1643e-10
n = 8800000000  error = 1.15132e-10
n = 8900000000  error = 1.13851e-10
n = 9000000000  error = 1.12596e-10
n = 9100000000  error = 1.11355e-10
n = 9200000000  error = 1.10163e-10
n = 9300000000  error = 1.08999e-10
n = 9400000000  error = 1.07854e-10
n = 9500000000  error = 1.0674e-10
n = 9600000000  error = 1.05633e-10
n = 9700000000  error = 1.04535e-10
n = 9800000000  error = 1.03481e-10

=====================================
n encontrado: 9858058109
π aproximado: 3.14159
Error: 9.99996e-11

**Resultado:**  
$$
n \approx 9.858 \times 10^{9} \text{ términos}
$$

El cálculo muestra que se necesitan aproximadamente $ 9{,}858{,}058{,}109 $ términos para alcanzar una precisión de $ 10^{-10} $, lo cual es computacionalmente muy costoso para la serie de arcotangente básica.

### 3. Fórmula de Machin para calcular $ \pi $

Otra fórmula para calcular $ \pi $ se puede deducir a partir de la identidad:

$$
\frac{\pi}{4} = 4 \arctan\left(\frac{1}{5}\right) - \arctan\left(\frac{1}{239}\right)
$$

Determine el número de términos que se deben sumar para garantizar una aproximación a $ \pi $ dentro de $ 10^{-3} $.


In [17]:
import math

tol = 1e-3

def arctan_series(x, n):
    s = 0
    for k in range(n):
        s += ((-1)**k) * (x**(2*k+1)) / (2*k+1)
    return s

n = 1
while True:
    pi_approx = 4 * (4 * arctan_series(1/5, n) - arctan_series(1/239, n))
    error = abs(math.pi - pi_approx)
    if error < tol:
        break
    n += 1

print("Número de términos necesarios:", n)
print("π aproximado:", pi_approx)
print("Error:", error)


Número de términos necesarios: 2
π aproximado: 3.1405970293260603
Error: 0.0009956242637327861


**Desarrollo:**

Usando la serie de Maclaurin para arcotangente en la fórmula de Machin:

$$
\pi \approx 16 \sum_{k=0}^{n-1} \frac{(-1)^k}{(2k+1)5^{2k+1}} - 4 \sum_{k=0}^{n-1} \frac{(-1)^k}{(2k+1)239^{2k+1}}
$$

**Resultado:**

Número de términos necesarios: $ n = 2 $

$ \pi $ aproximado: $ 3.1405970293260603 $

Error: $ 0.0009956242637327861 < 10^{-3} $ ✓

La fórmula de Machin converge mucho más rápido que la serie básica de $ \arctan(1) $, requiriendo solamente 2 términos en comparación con los 2000 términos necesarios en el ejercicio 2a.


### 4. Optimización de sumatorias

**Parte a)** ¿Cuántas multiplicaciones y adiciones se requieren para calcular $ \displaystyle \sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j $?

**Desarrollo:**

Para la forma directa $ \displaystyle \sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j $:

- **Multiplicaciones:** Para cada $ i $, se calculan $ n $ productos $ a_i b_j $. Total: $ n \times n = n^2 $ multiplicaciones.
- **Adiciones:** Para cada suma interna hay $ n-1 $ adiciones, y luego $ n-1 $ adiciones para sumar los $ n $ resultados. Total: $ n(n-1) + (n-1) = n^2 - 1 $ adiciones.

**Parte b)** Modifique la sumatoria del inciso a) a una forma equivalente que reduzca el número de cálculos.

**Desarrollo:**

La expresión se puede factorizar como:

$$
\sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j = \left(\sum_{i=1}^{n} a_i\right) \left(\sum_{j=1}^{n} b_j\right)
$$

Para esta forma optimizada:

- **Multiplicaciones:** Solo 1 multiplicación (producto de las dos sumas).
- **Adiciones:** $ (n-1) $ para $ \sum a_i $ y $ (n-1) $ para $ \sum b_j $. Total: $ 2n - 2 $ adiciones.

**Reducción de cálculos:**

- Multiplicaciones: de $ n^2 $ a $ 1 $
- Adiciones: de $ n^2 - 1 $ a $ 2n - 2 $


In [20]:

def count_multiplications(n: int):
    if n <= 0:
        return 0, 0
    count = n * (n + 1) // 2
    last_product = n * n   
    return count, last_product

n = 3
multiplicaciones, multiplicacion = count_multiplications(n)
multiplicaciones

6

## DISCUSIONES

### 1. Algoritmo para sumar la serie finita en orden inverso

Escriba un algoritmo para sumar la serie finita $ \displaystyle \sum_{i=1}^{n} x_i $ en orden inverso.

**Algoritmo:**

**ENTRADA:** $ n $, $ x_1, x_2, \ldots, x_n $  
**SALIDA:** $ \text{SUMA} $

**Paso 1:** Determine $ \text{SUMA} = 0 $

**Paso 2:** Para $ i = n, n-1, \ldots, 1 $ haga  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determine $ \text{SUMA} = \text{SUMA} + x_i $

**Paso 3:** SALIDA $ \text{SUMA} $; PARE.

**Implementación:**


In [24]:
def sum_back(iterations: int, i: int):
    result = 0
    for a in range(iterations, i-1, -1):

        result += a
    return result

In [26]:
sum_back(10, 1)

55

**Resultado:**

Para $ n = 10 $ y $ x_i = i $, la suma en orden inverso es:

$$
\sum_{i=1}^{10} i = 10 + 9 + 8 + \cdots + 1 = 55
$$


### 2. Convergencia de serie infinita

Suponga que

$$
\frac{1-2x}{1-x+x^2} + \frac{2x-4x^3}{1-x^2+x^4} + \frac{4x^3-8x^7}{1-x^4+x^8} + \cdots = \frac{1+2x}{1+x+x^2}
$$

para $ x < 1 $ y si $ x = 0.25 $. Escriba y ejecute un algoritmo que determine el número de términos necesarios en el lado izquierdo de la ecuación de tal forma que el lado izquierdo difiera del lado derecho en menos de $ 10^{-6} $.

**Implementación:**


In [27]:

x = 0.25
tol = 1e-6
s = 0.0
k = 0
target = (1 + 2 * x) / (1 + x + x**2)
max_k = 1000
while k < max_k:
    a = 2**k - 1
    b = 2**(k+1) - 1
    num = (2**k) * (x**a) - (2**(k+1)) * (x**b)
    den = 1 - x**(2**k) + x**(2**(k+1))
    s += num / den
    err = abs(s - target)
    if err < tol:
        break
    k += 1
print(k + 1, s, err)

4 1.1428571279559818 1.4901160971803051e-08


**Desarrollo:**

El término general de la serie tiene la forma:

$$
\frac{2^{k}x^{2^{k}-1} - 2^{k+1}x^{2^{k+1}-1}}{1 - x^{2^{k}} + x^{2^{k+1}}}
$$

donde $ k = 0, 1, 2, \ldots $

**Resultado:**

Para $ x = 0.25 $ y tolerancia $ 10^{-6} $:

- Número de términos necesarios: $ 4 $
- Suma obtenida: $ 1.1428571279559818 $
- Valor exacto: $ \frac{1+2(0.25)}{1+0.25+0.25^2} = \frac{1.5}{1.3125} \approx 1.142857142857 $
- Error: $ 1.49 \times 10^{-8} < 10^{-6} $ ✓

La serie converge rápidamente debido a que los términos decrecen de manera exponencial conforme $ x^{2^k} $ se aproxima a cero.


> **Nota** declaro el uso de IA(modelo GPT5) para explicacion de algunos ejercicios y para ayuda con markdown  y generacion de codigo
